GetVolumePathNamesForVolumeName failed with Win32 error code 433

Looks like ``Microsoft.Build.Tasks.Copy`` failed when there are some disconnected drives with have drive letter mapping.

```
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061: The "Copy" task could not be instantiated from "E:\packages\nuget\microsoft.build.copyonwrite\1.0.315\build\netstandard2.0\Microsoft.Build.CopyOnWrite.dll". 
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061: System.TypeInitializationException: The type initializer for 'Microsoft.Build.Tasks.Copy' threw an exception. ---> System.ComponentModel.Win32Exception: GetVolumePathNamesForVolumeName(\\?\Volume{55eecc11-0000-0000-0002-000000000000}\) failed with Win32 error code 433
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.CopyOnWrite.Windows.VolumeEnumerator.GetVolumePathNamesForVolumeName(String volumeName, Dictionary`2 driveLetterUpperToSubstDriveLettersUpper) in D:\CoW\lib\Windows\VolumeEnumerator.cs:line 128
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.CopyOnWrite.Windows.VolumeEnumerator.<GetVolumesAndVolumePaths>d__2.MoveNext() in D:\CoW\lib\Windows\VolumeEnumerator.cs:line 120
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.CopyOnWrite.Windows.VolumeInfoCache.BuildFromCurrentFilesystem() in D:\CoW\lib\Windows\VolumeInfoCache.cs:line 34
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.CopyOnWrite.CopyOnWriteFilesystemFactory.Create() in D:\CoW\lib\CopyOnWriteFilesystemFactory.cs:line 48
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at System.Lazy`1.CreateValue()
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at System.Lazy`1.LazyInitValue()
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.Build.Tasks.Copy..cctor()
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    --- End of inner exception stack trace ---
1>C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5034,5): error MSB4061:    at Microsoft.Build.Tasks.Copy..ctor()
```

 mountvol.exe gives me this output:

```
Possible values for VolumeName along with current mount points are:

    \\?\Volume{c798f43e-be3d-46a4-a0ed-4398088b775d}\
        I:\

    \\?\Volume{42165081-1a13-4bff-b2bf-2bd6b39af7e6}\
        C:\

    \\?\Volume{530e87c3-1667-43ff-b3cb-880ba5f4ac0b}\
        *** NO MOUNT POINTS ***

    \\?\Volume{2473cbb0-dcb5-42ac-b360-72e29db43828}\
        E:\

    \\?\Volume{18c7dd90-7ca1-4c6a-b58d-7949b027cbdc}\
        D:\

    \\?\Volume{a3f3fd35-9d09-476d-9954-ef650995739a}\
        *** NO MOUNT POINTS ***

    \\?\Volume{55eecc11-0000-0000-0002-000000000000}\
A device which does not exist was specified.
    \\?\Volume{d7a4bec3-0000-0000-0002-000000000000}\
A device which does not exist was specified.
```

I believe that the task should ignore that drives. CoW target was on 2473cbb0-dcb5-42ac-b360-72e29db43828 drive

--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -26,7 +26,7 @@
   <PropertyGroup>
 
     <!-- DOCSYNC: When changing version number update README.md -->
-    <Version>0.3.8.0</Version>
+    <Version>0.3.9.0</Version>
     <AssemblyVersion>0.9.9999.0</AssemblyVersion>
 
     <Company>Microsoft</Company>

--- a/README.md
+++ b/README.md
@@ -43,6 +43,7 @@ File clones on Windows do not actually allocate space on-drive for the clone. Th
 
 [![NuGet version (CopyOnWrite)](https://img.shields.io/nuget/v/CopyOnWrite?style=plastic)](https://www.nuget.org/packages/CopyOnWrite)
 
+* 0.3.9 September 2024: Fix https://github.com/microsoft/CopyOnWrite/issues/44 - follow up on ignoring FILE_NOT_FOUND on volume enumeration
 * 0.3.8 March 2024: Fix https://github.com/microsoft/MSBuildSdks/issues/546 - ignore FILE_NOT_FOUND on volume enumeration. Plus add SourceLink to the main library.
 * 0.3.7 September 2023: Fix #30 - ignore ACCESS_DENIED on volume enumeration to avoid need to escalate privilege on Windows.
 * 0.3.6 July 2023: Set AssemblyVersion to 0.9.9999.0 to allow mixing different minor-version binaries from different packages in the same appdomain/process.

--- a/lib/Windows/VolumeEnumerator.cs
+++ b/lib/Windows/VolumeEnumerator.cs
@@ -117,7 +117,13 @@ internal sealed class VolumeEnumerator : IDisposable
 
         while (GetNextVolume(out string? volumeName))
         {
-            yield return new VolumePaths(volumeName!, GetVolumePathNamesForVolumeName(volumeName!, driveLetterUpperToSubstDriveLetterUpper));
+            IReadOnlyList<string> volumePathNames = GetVolumePathNamesForVolumeName(volumeName!, driveLetterUpperToSubstDriveLetterUpper);
+            if (volumePathNames.Count == 0)
+            {
+                // Skip volumes with no mount points.
+                continue;
+            }
+            yield return new VolumePaths(volumeName!, volumePathNames);
         }
     }
 
@@ -137,6 +143,12 @@ internal sealed class VolumeEnumerator : IDisposable
             if (!success)
             {
                 int lastErr = Marshal.GetLastWin32Error();
+                if (lastErr == NativeMethods.ERROR_FILE_NOT_FOUND)
+                {
+                    // No mount points for this volume.
+                    return Array.Empty<string>();
+                }
+
                 if (lastErr != NativeMethods.ERROR_MORE_DATA)
                 {
                     throw new Win32Exception(lastErr,

--- a/lib/Windows/VolumeInfoCache.cs
+++ b/lib/Windows/VolumeInfoCache.cs
@@ -103,8 +103,6 @@ internal sealed class VolumeInfoCache
                                     "If the drive was added recently you may need to recreate the filesystem cache.");
     }
 
-    private const int ERROR_FILE_NOT_FOUND = 2;
-    private const int ERROR_ACCESS_DENIED = 5;
     private const int ERROR_NOT_READY = 21;
     private const int ERROR_INVALID_PARAMETER = 87;
     private const int ERROR_UNRECOGNIZED_VOLUME = 1005;
@@ -135,8 +133,8 @@ internal sealed class VolumeInfoCache
                 lastErr == ERROR_NOT_READY ||
                 lastErr == ERROR_INVALID_PARAMETER ||
                 lastErr == FVE_E_LOCKED_VOLUME ||
-                lastErr == ERROR_ACCESS_DENIED ||
-                lastErr == ERROR_FILE_NOT_FOUND)
+                lastErr == NativeMethods.ERROR_ACCESS_DENIED ||
+                lastErr == NativeMethods.ERROR_FILE_NOT_FOUND)
             {
                 return null;
             }

